Fixes #10710: Use instances of Mapping rather than dict in subs

parnmatt opened the pull request:
Fixes #10710. All standard library `dict`-like objects are instances of `Mapping` thus, `OrderedDict`, `ChainMap`, etc., can now be a drop-in replacement for a `dict` in substitutions. Rather than having to wrap the object in a `dict` call.

In Python 3.3 `collections.abc` separated, however to keep compatibility with `sys.version_info < (3,3)`
`from collections import Mapping` works for newer and older versions of Python.

parnmatt committed:
Use instances of Mapping rather than dict in subs

Fixes #10710. All standard library `dict`-like objects are instances of `Mapping` thus, `OrderedDict`, `ChainMap`, etc., can now be a drop-in replacement for a `dict` in substitutions. Rather than having to wrap the object in a `dict` call.

In Python 3.3 `collections.abc` separated, however to keep compatibility with `sys.version_info < (3,3)`
`from collections import Mapping` works for newer and older versions of Python.

moorepants commented:
Please add a small unit test that shows this working with an OrderedDict or ChainMap.

parnmatt commented:
I cannot see a place where the unit tests are stored.
If I know where to write one, I will, shouldn't take long.

parnmatt committed:
Add tests for #10753

Added testing for substituions for `collections.ChainMap` and `collections.OrderedDict`

moorepants commented:
That looks good. Once tests pass, I'm +1.

parnmatt commented:
I've just realised that `OrderedDict` was introduced in 2.7, and `ChainMap` was introduced in 3.3
Thus these tests would need conditionals on version for the import and the asserts.
I'll do that tomorrow.

parnmatt committed:
Correct python version specific objects

`collections.OrderedDict` was introduced in Python 2.7, and `collections.ChainMap` was introduced in Python 3.3.

parnmatt commented:
All tests have passed, except the Python 3.5 ones, however I am having trouble trying to figure out what went wrong with the limited Travis log shown.

smichr commented:
Tests passed so I will commit as is.
